### PR TITLE
[WIP] [React]: Speed-dials

### PR DESCRIPTION
**Pull request by giota-cliqz**

### CODE REVIEW

File: packages/cliqz-home-react/package.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
   },
   "devDependencies": {
     "react-scripts": "0.9.5",
+    "react-toggle-display": "^2.2.0",
```
Comment (chrmod) [inline packages/cliqz-home-react/package.json:11]: that is rather a regular, not a dev dependency (same for spanan)

File: packages/cliqz-home-react/src/AddSpeedDial.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
+import React, { Component } from 'react';
```
Comment (chrmod) [inline packages/cliqz-home-react/src/AddSpeedDial.js:1]: would rather a call a component with a noun, not a verb, for example `new-speed-dial-form`

File: packages/cliqz-home-react/src/AddSpeedDial.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+      
+    })
+  }
+
+  render() {
+    return (
+      <div>
+        <ToggleDisplay show={this.state.show}>
+          <a href="#" onClick={this.handleClick}>
+            <svg width="40px" height="40px" viewBox="0 0 40 40" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlnsXlink="http://www.w3.org/1999/xlink">
```
Comment (chrmod) [inline packages/cliqz-home-react/src/AddSpeedDial.js:68]: shouldn't that be an image tag?

File: packages/cliqz-home-react/src/SpeedDialsRow.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+    // this.closeUndo = this.closeUndo.bind(this);
+    this.undoRemoval = this.undoRemoval.bind(this);
+    this.resetAll = this.resetAll.bind(this);
+  }
+
+  componentWillMount() {
+    this.state = {
+      isCustom: this.props.type === 'custom',
+      showAddButton: () => {
```
Comment (chrmod) [inline packages/cliqz-home-react/src/SpeedDialsRow.js:18]: should we put functions on a state?

File: packages/cliqz-home-react/src/SpeedDialsRow.js [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+      type: 'home',
+      action: 'click',
+      target_type: 'close-' + type + '-undo'
+    });
+  }
+
+  _reAddSpeedDial(dial) {
+    const index = dial.removedAt;
+    this.props.addSpeedDial(dial, index);
+    cliqz.freshtab.addSpeedDial(dial.url, index).then(() => {
```
Comment (chrmod) [inline packages/cliqz-home-react/src/SpeedDialsRow.js:59]: not super sure about it, but those two `addSpeedDial` functions should probably be merged

File: packages/cliqz-home-react/src/SpeedDialsRow.js [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+                 onClick={this.undoRemoval}>
+                 Undo
+              </a>
+              <a className="resetAll"
+                 href="#"
+                 onClick={this.resetAll}>
+                 Reset all
+              </a>
+              <a className="closeUndo"
+                 onClick={this.closeUndo.bind(this, this.props.type)}>
```
Comment (chrmod) [inline packages/cliqz-home-react/src/SpeedDialsRow.js:137]: wonder which patter in nicer, this or `onClick={() => this.closeUndo(this.props.type)}` or rather `onClick={() => this.onClick()}`

the last one is the abbervation of `onClick={this.onClick}`, assuming `onClick` is bound to this - which opens a question if we should use bind in JSX components

File: packages/cliqz-home-react/src/SpeedDialsRow.js [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    cliqz.core.sendTelemetry({
+      type: 'home',
+      action: 'click',
+      target_type: 'undo-' + type + '-remove'
+    });
+  }
+
+  resetAll() {
+    cliqz.freshtab.resetAllHistory().then( results => {
+      this.closeUndo.call(this, 'history');
```
Comment (chrmod) [inline packages/cliqz-home-react/src/SpeedDialsRow.js:89]: no need for `call`

File: packages/cliqz-home-react/src/SpeedDialsRow.js [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    const type = speedDial.custom ? 'custom' : 'history';
+
+    if(speedDial.custom) {
+      this._reAddSpeedDial.call(this, speedDial);
+    } else {
+      cliqz.freshtab.revertHistorySpeedDial(speedDial.url);
+      this.props.addSpeedDial(speedDial, speedDial.removedAt);
+    }
+
+    cliqz.core.sendTelemetry({
```
Comment (chrmod) [inline packages/cliqz-home-react/src/SpeedDialsRow.js:80]: how about putting all telemetry signals in one file (as we did in `dropdown`) to hide their complexity a little bit?

File: packages/cliqz-home-react/src/SpeedDialsRow.js [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+      });
+      this.setState({
+        removedItems: newItems
+      });
+    })
+  }
+
+  undoRemoval() {
+    const speedDial = this.state.removedItems.pop();
+    const type = speedDial.custom ? 'custom' : 'history';
```
Comment (chrmod) [inline packages/cliqz-home-react/src/SpeedDialsRow.js:71]: `type` could be a getter on `SpeedDial` model

File: packages/cliqz-home-react/src/SpeedDialsRow.js [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        return this.props.dials.length < 5;
+      },
+      removedItems: []
+    }
+  }
+
+  removeSpeedDial(dial, index) {
+    this.props.removeSpeedDial(dial, index);
+    dial.removedAt = index;
+    this.state.removedItems.push(dial);
```
Comment (chrmod) [inline packages/cliqz-home-react/src/SpeedDialsRow.js:31]: not sure if we should change a state like this. this is probably more idiomatic:
```
this.setState({
  removedItems: [
    ...this.state.removedItems,
    dial,
  ],
});

File: packages/cliqz-home-react/src/SpeedDialsRow.js [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    super(props);
+
+    // this.closeUndo = this.closeUndo.bind(this);
+    this.undoRemoval = this.undoRemoval.bind(this);
+    this.resetAll = this.resetAll.bind(this);
+  }
+
+  componentWillMount() {
+    this.state = {
+      isCustom: this.props.type === 'custom',
```
Comment (chrmod) [inline packages/cliqz-home-react/src/SpeedDialsRow.js:17]: we some times use `isCustom` and sometimes a `type` would be nice if we could clean it up on this occasion

File: packages/cliqz-home-react/src/SpeedDial.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
+import React, { Component } from 'react';
+import Logo from './Logo';
+
+const styles = {
+  position: 'relative'
+}
+
+class SpeedDial extends Component {
+  constructor(props) {
```
Comment (chrmod) [inline packages/cliqz-home-react/src/SpeedDial.js:9]: no need for constructor is it sets nothing

File: packages/cliqz-home-react/src/App.js [hunk 13]
```diff
@@ -44,6 +48,39 @@ class App extends Component {
...
+    const isCustom = dial.custom;
+    let dialType = isCustom ? 'custom' : 'history';
+    let newItems = this.state.dials[dialType].filter((item) => {
+      return item != dial;
+    });
+
+    var obj = this.state.dials;
+    obj[dialType] = newItems;
+   
+    this.setState({
```
Comment (chrmod) [inline packages/cliqz-home-react/src/App.js:77]: how about:
```
this.setState({
  dials: {
    ...this.state.dials,
    [dialType]: this.state.dials.filter(d => d !== dial),
  },
});
```

**Review by chrmod — CHANGES_REQUESTED** [review]
bunch of stuff to discuss, but over all good job!

**Comment by chrmod** [comment]
@mai-cliqz can you please take a look on this one?

File: packages/cliqz-home-react/src/AddSpeedDial.js (hunk 3 above)
Comment (mai-cliqz) [reply packages/cliqz-home-react/src/AddSpeedDial.js:68]: We could use svg tag if we want to change the fill or stroke of the image programmatically.

File: packages/cliqz-home-react/src/App.js [hunk 14]
```diff
@@ -62,7 +99,17 @@ class App extends Component {
           <nav id="nav-left"></nav>
           <section id="content">
             <section id="top">
-              <SpeedDials dials={this.state.dials} />
+              <SpeedDialsRow 
+                dials={this.state.dials.history} 
+                type="history"
+                removeSpeedDial={this.removeSpeedDial}
+                addSpeedDial={this.addSpeedDial}
+                getSpeedDials={this.getSpeedDials} />
```
Comment (mai-cliqz) [inline packages/cliqz-home-react/src/App.js:107]: Should we delegate these functions to speed dial component ?

**Review by mai-cliqz — COMMENTED** [review]
@giota-cliqz So internalization will go to next PR ?

**Comment by chrmod** [comment]
@mai-cliqz internationalization will come through `chrome.i18n`, which I will implement for out bootstrap addon.

**Comment by chrmod** [comment]
For sake of tomorrows merging to browser-core it got merged here

**Pull request merged**